js/src/modals/PinMatrix/pinMatrix.js

```diff
@@ -0,0 +1,147 @@
+    device: PropTypes.object.isRequired
+  }
+
+  constructor () {
+    super();
+
+    this.state = {
+      passcode: '',
+      failureMessage: ''
+    };
+
+    this.pinMatrix = [7, 8, 9, 4, 5, 6, 1, 2, 3];
+  }
+
+  render () {
+    const { failureMessage, passcode } = this.state;
+    const { device } = this.props;
+
+    return (
+      <div className={ styles.pinMatrix }>
+
+        <div className={ styles.overlay } />
+
+        <div className={ styles.modal }>
+          <div className={ styles.body }>
+            <div id={ styles.title }>
+              <span>
+                <FormattedMessage
+                  id='pinMatrix.enterPin'
+                  defaultMessage='Please enter the pin for your {manufacturer} hardware wallet'
+                  values={ {
+                    manufacturer: device.manufacturer
+                  } }
+                />
+              </span>
+            </div>
+            <div id={ styles.passcodeBoxes }>
+              {this.renderPasscodeBox()}
+            </div>
+
+            <div id={ styles.pin }>
+              {passcode.replace(/./g, '*')}
+              {
+                (passcode.length)
+                  ? <div id={ styles.clearThik } onClick={ this.removeDigit } />
+                  : null
+              }
+            </div>
+            <div>
+              <span className={ styles.button } id={ styles.submit } onClick={ this.submit }>Submit</span>
+            </div>
+            <div id={ styles.error }>
+              { failureMessage }
+            </div>
+          </div>
+        </div>
+
+      </div>
+    );
+  }
+
+  renderPasscodeBox () {
+    return Array.apply(null, Array(9)).map((box, index) => {
+      let addDigit = () => this.addDigit(this.pinMatrix[index]);
+
+      return (
+        <div
+          className={ styles.passcodeBox }
+          onClick={ addDigit }
+          key={ index }
+        >
+          <div className={ styles.passcodeBall } />
+        </div>
+      );
+    });
+  }
+
+  addDigit = (digit) => {
+    if (this.state.passcode.length > 8) {
+      return;
+    }
+    this.setState({
+      passcode: this.state.passcode + digit
+    });
+  }
+
+  removeDigit = () => {
+    this.setState({
+      passcode: this.state.passcode.slice(0, -1)
+    });
+  }
+
+  clearPasscode = () => {
+    this.setState({
+      passcode: ''
+    });
+  }
+
+  submit = () => {
+    const { device, store } = this.props;
+    const { passcode } = this.state;
+
+    store.pinMatrixAck(device, passcode)
+      .then((status) => {
+        if (!status) {
+          this.setState({
+            passcode: '',
+            failureMessage: (
+              <FormattedMessage
+                id='pinMatrix.label.failureMessage'
+                defaultMessage='Wrong pin, try again.'
+              />
+            )
+          });
+        } else {
+          this.setState({
+            passcode: '',
+            failureMessage: ''
+          });
+        }
+      });
```

Comment:
What if it fails?

Reply:
Yeah, this should really be a message like "Failed to communicate with device", a failure here means it wasn't able to even send the pin to the device.